File: snippets/rust/src/non_bitcoin_asset.rs
use std::sync::Arc;

use anyhow::Result;
use breez_sdk_liquid::prelude::*;
use log::info;

async fn prepare_receive_asset(sdk: Arc<LiquidSdk>) -> Result<()> {
    // ANCHOR: prepare-receive-payment-asset
    // Create a Liquid BIP21 URI/address to receive an asset payment to.
    // Note: Not setting the amount will generate an amountless BIP21 URI.
    let usdt_asset_id =
        "ce091c998b83c78bb71a632313ba3760f1763d9cfcffae02258ffa9865a37bd2".to_string();
    let optional_amount = Some(ReceiveAmount::Asset {
        asset_id: usdt_asset_id,
        payer_amount: Some(1.50),
    });
    let prepare_response = sdk
        .prepare_receive_payment(&PrepareReceiveRequest {
            payment_method: PaymentMethod::LiquidAddress,
            amount: optional_amount,
        })
        .await?;

    // If the fees are acceptable, continue to create the Receive Payment
    let receive_fees_sat = prepare_response.fees_sat;
    info!("Fees: {} sats", receive_fees_sat);
    // ANCHOR_END: prepare-receive-payment-asset
    Ok(())
}

async fn prepare_send_payment_asset(sdk: Arc<LiquidSdk>) -> Result<()> {
    // ANCHOR: prepare-send-payment-asset
    // Set the Liquid BIP21 or Liquid address you wish to pay
    let destination = "<Liquid BIP21 or address>".to_string();
    // If the destination is an address or an amountless BIP21 URI,
    // you must specify an asset amount
    let usdt_asset_id =
        "ce091c998b83c78bb71a632313ba3760f1763d9cfcffae02258ffa9865a37bd2".to_string();
    let optional_amount = Some(PayAmount::Asset {
        to_asset: usdt_asset_id,
        from_asset: None,
        receiver_amount: 1.50,
        estimate_asset_fees: Some(false),
    });
    let prepare_response = sdk
        .prepare_send_payment(&PrepareSendRequest {
            destination,
            amount: optional_amount,
            disable_mrh: None,
            payment_timeout_sec: None,
        })
        .await?;

    // If the fees are acceptable, continue to create the Send Payment
    let send_fees_sat = prepare_response.fees_sat;
    info!("Fees: {:?} sats", send_fees_sat);
    // ANCHOR_END: prepare-send-payment-asset
    Ok(())
}

async fn prepare_send_payment_asset_fees(sdk: Arc<LiquidSdk>) -> Result<()> {
    // ANCHOR: prepare-send-payment-asset-fees
    let destination = "<Liquid BIP21 or address>".to_string();
    let usdt_asset_id =
        "ce091c998b83c78bb71a632313ba3760f1763d9cfcffae02258ffa9865a37bd2".to_string();
    // Set the optional estimate asset fees param to true
    let optional_amount = Some(PayAmount::Asset {
        to_asset: usdt_asset_id,
        from_asset: None,
        receiver_amount: 1.50,
        estimate_asset_fees: Some(true),
    });
    let prepare_response = sdk
        .prepare_send_payment(&PrepareSendRequest {
            destination,
            amount: optional_amount,
            disable_mrh: None,
            payment_timeout_sec: None,
        })
        .await?;

    // If the asset fees are set, you can use these fees to pay to send the asset
    let send_asset_fees = prepare_response.estimated_asset_fees;
    info!("Estimated Fees: ~{:?}", send_asset_fees);

    // If the asset fess are not set, you can use the sats fees to pay to send the asset
    let send_fees_sat = prepare_response.fees_sat;
    info!("Fees: {:?} sats", send_fees_sat);
    // ANCHOR_END: prepare-send-payment-asset-fees
    Ok(())
}

async fn send_self_payment_asset(sdk: Arc<LiquidSdk>) -> Result<()> {
    // ANCHOR: send-self-payment-asset
    // Create a Liquid address to receive to
    let prepare_receive_res = sdk
        .prepare_receive_payment(&PrepareReceiveRequest {
            payment_method: PaymentMethod::LiquidAddress,
            amount: None,
        })
        .await?;
    let receive_res = sdk
        .receive_payment(&ReceivePaymentRequest {
            prepare_response: prepare_receive_res,
            description: None,
            use_description_hash: None,
            payer_note: None,
        })
        .await?;

    // Swap your funds to the address we've created
    let usdt_asset_id =
        "ce091c998b83c78bb71a632313ba3760f1763d9cfcffae02258ffa9865a37bd2".to_string();
    let btc_asset_id =
        "6f0279e9ed041c3d710a9f57d0c02928416460c4b722ae3457a11eec381c526d".to_string();
    let prepare_send_res = sdk
        .prepare_send_payment(&PrepareSendRequest {
            destination: receive_res.destination,
            amount: Some(PayAmount::Asset {
                to_asset: usdt_asset_id,
                // We want to receive 1.5 USDt
                receiver_amount: 1.5,
                estimate_asset_fees: None,
                from_asset: Some(btc_asset_id),
            }),
            disable_mrh: None,
            payment_timeout_sec: None,
        })
        .await?;
    let payment = sdk
        .send_payment(&SendPaymentRequest {
            prepare_response: prepare_send_res,
            use_asset_fees: None,
            payer_note: None,
        })
        .await?;
    // ANCHOR_END: send-self-payment-asset
    dbg!(payment);
    Ok(())
}

async fn send_payment_fees(
    sdk: Arc<LiquidSdk>,
    prepare_response: PrepareSendResponse,
) -> Result<()> {
    // ANCHOR: send-payment-fees
    // Set the use asset fees param to true
    let send_response = sdk
        .send_payment(&SendPaymentRequest {
            prepare_response,
            payer_note: None,
            use_asset_fees: Some(true),
        })
        .await?;
    let payment = send_response.payment;
    // ANCHOR_END: send-payment-fees
    dbg!(payment);
    Ok(())
}

async fn fetch_asset_balance(sdk: Arc<LiquidSdk>) -> Result<()> {
    // ANCHOR: fetch-asset-balance
    let info = sdk.get_info().await?;
    let asset_balances = info.wallet_info.asset_balances;
    // ANCHOR_END: fetch-asset-balance

    Ok(())
}

File: snippets/go/non_bitcoin_asset.go
package example

import (
	"log"

	"github.com/breez/breez-sdk-liquid-go/breez_sdk_liquid"
)

func PrepareReceiveAsset(sdk *breez_sdk_liquid.BindingLiquidSdk) {
	// ANCHOR: prepare-receive-payment-asset
	// Create a Liquid BIP21 URI/address to receive an asset payment to.
	// Note: Not setting the amount will generate an amountless BIP21 URI.
	usdtAssetId := "ce091c998b83c78bb71a632313ba3760f1763d9cfcffae02258ffa9865a37bd2"
	payerAmount := float64(1.50)
	var optionalAmount breez_sdk_liquid.ReceiveAmount = breez_sdk_liquid.ReceiveAmountAsset{
		AssetId:     usdtAssetId,
		PayerAmount: &payerAmount,
	}
	prepareRequest := breez_sdk_liquid.PrepareReceiveRequest{
		PaymentMethod: breez_sdk_liquid.PaymentMethodLiquidAddress,
		Amount:        &optionalAmount,
	}
	if prepareResponse, err := sdk.PrepareReceivePayment(prepareRequest); err == nil {
		// If the fees are acceptable, continue to create the Receive Payment
		receiveFeesSat := prepareResponse.FeesSat
		log.Printf("Fees: %v sats", receiveFeesSat)
	}
	// ANCHOR_END: prepare-receive-payment-asset
}

func PrepareSendPaymentAsset(sdk *breez_sdk_liquid.BindingLiquidSdk) {
	// ANCHOR: prepare-send-payment-asset
	// Set the Liquid BIP21 or Liquid address you wish to pay
	destination := "<Liquid BIP21 or address>"
	// If the destination is an address or an amountless BIP21 URI,
	// you must specify an asset amount
	usdtAssetId := "ce091c998b83c78bb71a632313ba3760f1763d9cfcffae02258ffa9865a37bd2"
	receiverAmount := float64(1.50)
	estimateAssetFees := false
	var optionalAmount breez_sdk_liquid.PayAmount = breez_sdk_liquid.PayAmountAsset{
		AssetId:           usdtAssetId,
		ReceiverAmount:    receiverAmount,
		EstimateAssetFees: &estimateAssetFees,
	}

	prepareRequest := breez_sdk_liquid.PrepareSendRequest{
		Destination: destination,
		Amount:      &optionalAmount,
	}
	prepareResponse, err := sdk.PrepareSendPayment(prepareRequest)
	if err != nil {
		log.Printf("Error: %#v", err)
		return
	}

	sendFeesSat := prepareResponse.FeesSat
	log.Printf("Fees: %v sats", sendFeesSat)
	// ANCHOR_END: prepare-send-payment-asset
}

func PrepareSendPaymentAssetFees(sdk *breez_sdk_liquid.BindingLiquidSdk) {
	// ANCHOR: prepare-send-payment-asset-fees
	destination := "<Liquid BIP21 or address>"
	usdtAssetId := "ce091c998b83c78bb71a632313ba3760f1763d9cfcffae02258ffa9865a37bd2"
	receiverAmount := float64(1.50)
	// Set the optional estimate asset fees param to true
	estimateAssetFees := true
	var optionalAmount breez_sdk_liquid.PayAmount = breez_sdk_liquid.PayAmountAsset{
		AssetId:           usdtAssetId,
		ReceiverAmount:    receiverAmount,
		EstimateAssetFees: &estimateAssetFees,
	}

	prepareRequest := breez_sdk_liquid.PrepareSendRequest{
		Destination: destination,
		Amount:      &optionalAmount,
	}
	prepareResponse, err := sdk.PrepareSendPayment(prepareRequest)
	if err != nil {
		log.Printf("Error: %#v", err)
		return
	}

	// If the asset fees are set, you can use these fees to pay to send the asset
	sendAssetFees := prepareResponse.EstimatedAssetFees
	log.Printf("Estimated Fees: ~%v", sendAssetFees)

	// If the asset fess are not set, you can use the sats fees to pay to send the asset
	sendFeesSat := prepareResponse.FeesSat
	log.Printf("Fees: %v sats", sendFeesSat)
	// ANCHOR_END: prepare-send-payment-asset-fees
}

func SendPaymentFees(sdk *breez_sdk_liquid.BindingLiquidSdk, prepareResponse breez_sdk_liquid.PrepareSendResponse) {
	// ANCHOR: send-payment-fees
	// Set the use asset fees param to true
	useAssetFees := true
	req := breez_sdk_liquid.SendPaymentRequest{
		PrepareResponse: prepareResponse,
		UseAssetFees:    &useAssetFees,
	}
	if response, err := sdk.SendPayment(req); err == nil {
		payment := response.Payment
		log.Printf("Payment: %#v", payment)
	}
	// ANCHOR_END: send-payment-fees
}

func ConfigureAssetMetadata() error {
	// ANCHOR: configure-asset-metadata
	// Create the default config
	breezApiKey := "<your-Breez-API-key>"
	config, err := breez_sdk_liquid.DefaultConfig(breez_sdk_liquid.LiquidNetworkMainnet, &breezApiKey)
	if err != nil {
		return err
	}

	// Configure asset metadata. Setting the optional fiat ID will enable
	// paying fees using the asset (if available).
	fiatId := "EUR"
	assetMetadata := []breez_sdk_liquid.AssetMetadata{
		{
			AssetId:   "18729918ab4bca843656f08d4dd877bed6641fbd596a0a963abbf199cfeb3cec",
			Name:      "PEGx EUR",
			Ticker:    "EURx",
			Precision: 8,
			FiatId:    &fiatId,
		},
	}
	config.AssetMetadata = &assetMetadata
	// ANCHOR_END: configure-asset-metadata
	return nil
}

func FetchAssetBalance(sdk *breez_sdk_liquid.BindingLiquidSdk) {
	// ANCHOR: fetch-asset-balance
	if info, err := sdk.GetInfo(); err == nil {
		assetBalances := info.WalletInfo.AssetBalances
		log.Printf("Asset balances: %#v", assetBalances)
	}
	// ANCHOR_END: fetch-asset-balance
}
